File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        vector<int>pre(n,1);
        vector<int>suf(n,1);

        for(int i=1;i<n;i++)
        {
            pre[i] = pre[i-1]*nums[i-1];
        }
        for(int i=n-2;i>=0;i--)
        {
            suf[i] = suf[i+1]*nums[i+1];
        }
        vector<int>res;
        for(int i=0;i<n;i++)
        {
            res.push_back(pre[i]*suf[i]);
        }
        return res;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pref = 1
        suff = 1
        res = [1 for i in range(n)]
        for i in range(n):
            res[i] *= pref
            pref *= nums[i]
        for i in range(n-1,-1,-1):
            res[i] *= suff
            suff *= nums[i]
        return res
            